### ISSUE TITLE
Initial suggestions for plugin release

### ISSUE THREAD

**Issue by madhushreeray30** [gabrielsoltz/steampipe-plugin-semgrep#7]
Thanks @gabrielsoltz for this new plugin. Great work 🎉 !!

The basic structure looks good so far. While using the plugin, we did come up with a few suggestions based on our best practices:

`config/semgrep.spc`
  - Could you please update the `semgrep.spc` file to follow the format of the [namecheap plugin](https://github.com/turbot/steampipe-plugin-namecheap/blob/main/config/namecheap.spc)
  - Can you mention which of these parameters (`base_url` and `token`) are required and which are optional as done in the [namecheap plugin](https://github.com/turbot/steampipe-plugin-namecheap/blob/main/config/namecheap.spc)
  - We should mention values for both `base_url` and `token` however commented out and the values should be in the format of the actual values (however incorrect)
  - We should also add details to set these parameters via the environment variable
  - For instance:
```
connection "semgrep" {
  plugin = "gabrielsoltz/semgrep"
  
  # The base URL of Semgrep. Required.
  # This can be set via the `SEMGREP_URL` environment variable.
  # base_url = "https://semgrep.dev/api/v1"

  # The access token required for API calls. Required.
  # This can also be set via the `SEMGREP_TOKEN` environment variable.
  # token = "45f86adc2nv54efd76151530rr629fc8953c2a111111fd74fa7d361d70e55759"
}
```

`docs/tables/table_semgrep_deployments.md`

  - The name of the table should be in singular. We should change it to `semgrep_deployment`
  - We can give a better description of the table like:
```
Semgrep deployments refer to the process of implementing and using the Semgrep tool within a software development or security workflow. Semgrep deployments involve the strategic integration of the Semgrep static code analysis tool into a software development pipeline or security assessment process.
```
  - The formatting of the query is incorrect, we can use [free formatted](https://www.freeformatter.com/sql-formatter.html#before-output) (2 indent spaces) for formatting the queries
  - The example queries in the doc look a bit light. Could you please add 2 to 3 more example queries wherever possible?

`docs/tables/table_semgrep_findings.md`
 
  - The name of the table should be in singular. We should change it to `semgrep_finding`
  - We can give a better description of the table like:
```
A Semgrep finding is a detailed report or notification generated by the Semgrep static code analysis tool when it detects a problem or anomaly within the source code of a software application. These findings typically include information about the location of the issue within the codebase, a description of the problem, and often a severity rating to help prioritize and address the identified concern.
```
  - The formatting of the query is incorrect, we can use [free formatted](https://www.freeformatter.com/sql-formatter.html#before-output) (2 indent spaces) for formatting the queries
  - To be more specific we should update the title of the query from `List all Semgrep with high severity that are not triaged` to `List all Semgrep findings with high severity that are not triaged`
  - The example queries in the doc look a bit light. Could you please add 2 to 3 more example queries wherever possible?
 
`docs/tables/table_semgrep_projects.md`

  - The name of the table should be in singular. We should change it to `semgrep_project`
  - We can give a better description of the table like:
```
A Semgrep project is a focused and organized endeavor undertaken by a development or security team to leverage the capabilities of the Semgrep static code analysis tool. It involves the application of Semgrep to analyze and improve the quality, security, and compliance of one or more software projects.
```
  - The formatting of the query is incorrect, we can use [free formatted](https://www.freeformatter.com/sql-formatter.html#before-output) (2 indent spaces) for formatting the queries
  - The example queries in the doc look a bit light. Could you please add 2 to 3 more example queries wherever possible?

`docs/index.md`
  - The description: `Steampipe plugin to query Semgrep` can be modified to `Steampipe plugin to query deployments, findings, and projects from Semgrep`
  - The output of the example queries are missing the end line it should be added
  - One example query is enough for `index.md`
  - The table definition is incorrect, it should be>
   `**[Table definitions & examples →](/plugins/gabrielsoltz/semgrep/tables)**`
  - Also please make sure that the `Configuration section` is in sync with the changes suggested for the `config/semgrep.spc` file.
  - While mentioning about the environment variables it would be better to state the exact steps like :

Alternatively, you can also use the standard Semgrep environment variables to obtain credentials **only if other arguments (`base_url`and `token`) are not specified** in the connection:

```sh
export SEMGREP_URL=https://semgrep.dev/api/v1
export SEMGREP_TOKEN=45f86adc2nv54efd76151530rr629fc8953c2a111111fd74fa7d361d70e55759
``` 

`semgrep/connection_config.go`

  - Looks perfect 👍 
 
`semgrep/plugin.go`

  - The indentation is incorrect please use `Ctrl + S` or `Cmd + S` in case you are using a code editor like VS code
  - The list of tables should be sorted in ascending order
 
`semgrep/utils.go`
  - We should add the configuration for loading the connection from the cache, which ensures throttling protection, etc. Therefore we should also add the configuration to set it back to the cache once the connection is successfully created.
  
`semgrep/table_semgrep_deployments.go`

  - The file name and the table name should be changed to singular i.e `table_semgrep_deployment`
  - The table should have a description for instance `A table tracking Semgrep deployment details, including name and id.`
  - Each column must have a valid column description that can help us identify the purpose and usage of the column. The description should end with a `.`
  
`semgrep/table_semgrep_findings.go`

  - The file name and the table name should be changed to singular i.e `table_semgrep_finding`
  - The table should have a description for instance `Table for storing Semgrep findings data, including issues and metadata.`
  - Each column must have a valid column description that can help us identify the purpose and usage of the column. The description should end with a `.`
  - Instead of using `deployment_slug := d.EqualsQuals["deployment_slug"].GetStringValue()` we should use `deployment_slug := d.EqualsQualString("deployment_slug")`

`semgrep/table_semgrep_projects.go`

  - The file name and the table name should be changed to singular i.e `table_semgrep_finding`
  - The table should have a description for instance `Table for Semgrep projects, containing project-specific information and configurations.`
  - Each column must have a valid column description that can help us identify the purpose and usage of the column. The description should end with a `.`
  - Instead of using `deployment_slug := d.EqualsQuals["deployment_slug"].GetStringValue()` we should use `deployment_slug := d.EqualsQualString("deployment_slug")`

`CHANGELOG.md`

  - Please draft an initial Changelog for the plugin release, an example would be as follows:
```
## v0.0.1 [TBD]

_What's new?_

- New tables added
  - [semgrep_deployment](https://github.com/gabrielsoltz/steampipe-plugin-semgrep/blob/main/semgrep/table_semgrep_deployment.go)
  - [semgrep_finding](https://github.com/gabrielsoltz/steampipe-plugin-semgrep/blob/main/semgrep/table_semgrep_finding.go)
  - [semgrep_project](https://github.com/gabrielsoltz/steampipe-plugin-semgrep/blob/main/semgrep/table_semgrep_project.go)
```
Please note that the example contains the name in singular assuming that the changelog will be added after updating the table names.

`Makefile`

  - The file need not check for the operating system just the install command is enough
  
`README.md`

  - Please add the graphic image link at the top once it is ready from our end
  - For `Use SQL to query your security findings from Semgrep.` it would be helpful to add the link to the website i.e `Use SQL to query your security findings from [Semgrep](https://semgrep.dev/)`
  - The get started, and documentation links should be updated and the community and get involved sections should be added. For reference :
```
- **[Get started →](https://hub.steampipe.io/plugins/gabrielsoltz/semgrep)**
- Documentation: [Table definitions & examples](https://hub.steampipe.io/plugins/gabrielsoltz/semgrep/tables)
- Community: [Slack Channel](https://steampipe.io/community/join)
- Get involved: [Issues](https://github.com/gabrielsoltz/steampipe-plugin-semgrep/issues)
```
  - We have missed the configuration section, we should add it and it should be followed by the query:
  
```
Configure the server address in `~/.steampipe/config/semgrep.spc`:

```hcl
connection "tomba" {
  plugin = "gabrielsoltz/semgrep"

  # Authentication information
  base_url = https://semgrep.dev/api/v1
  token = "45f86adc2nv54efd76151530rr629fc8953c2a111111fd74fa7d361d70e55759"
}
```

Or through environment variables:

```sh
export SEMGREP_URL=https://semgrep.dev/api/v1
export SEMGREP_TOKEN=45f86adc2nv54efd76151530rr629fc8953c2a111111fd74fa7d361d70e55759
```

Run a query:

```sql
select 
  id, 
  state, 
  rule_message 
from 
  semgrep_findings 
where 
  state = 'unresolved' 
  and deployment_slug = 'my-company';
```

Note : We should add the output of the query here 

  - Please add the details of the help wanted issues at the end:
```
`help wanted` issues:

- [Steampipe](https://github.com/turbot/steampipe/labels/help%20wanted)
- [semgrep Plugin](https://github.com/gabrielsoltz/steampipe-plugin-semgrep/labels/help%20wanted)
```

`go.mod`
  - Please recompile the plugin with `steampipe-plugin-sdk v5.6.2` and then run `go mod tidy`

`Follow-up questions`

  - A couple of follow-up questions:
    - How are we handling the resource not found errors?
    - Do the APIs support pagination? I did not find any reference to it, but I want to make sure that we have checked that box before releasing it.
    - While working with the APIs did you encounter rate limit errors?
    - Please refer to https://steampipe.io/docs/develop/plugin-release-checklist#data-ingestion for more information

Please let us know if you have questions, happy to help 👍

**Comment by gabrielsoltz** [comment]
Hello @madhushreeray30, thank you so much for this review!

I applied all the mentioned suggestions here: #9 if you want to review it. 

The following points are pending; I'll create a new PR for each, but first, I need to investigate. 
- Cache
- Not Found errors
- Pagination
- Rate Limit

**Comment by gabrielsoltz** [comment]
Hello @madhushreeray30,

This is ready to be reviewed again.

- All syntaxis, linting, and documentation suggestions were implemented
- SDK updated 
- Pagination: I added support for every supported pagination API call
- Rate Limit: I didn't experience any rate limit issues while testing and working with the API. There is no mention in the documentation of any limitation, so I assume there is not. 
- Cache: As this is a simple GET request, I don't see how cache could be implemented and if it makes sense. Let me know if there is something I'm missing. 
- Not found errors: I'm not sure about this one. Queries return no elements when `were` clauses don't find resources.

**Comment by madhushreeray30** [comment]
Hey @jplanckeel thanks for making the suggested changes. Few suggestions after you have made the fixes, some were missed and some not addressed please can you have a look at these

- `semgrep.spc`
  - Please comment out the base_url value 
  ```#base_url = "https://semgrep.dev/api/v1"```
- `docs/tables`
  - Please format the queries with `2 indent spaces` with [free formatted](https://www.freeformatter.com/sql-formatter.html#before-output)
  - Please make sure that the descriptions end with a single `.` for instance in table semgrep_finding we get to see the column > 
  `{Name: "rule_name", Type: proto.ColumnType_STRING, Description: "Rule name of rule triggering.."},`
 please update the description to ```Description: "Rule name of rule triggering."```
  - Please can you add some more example queries for `semgrep_finding` table
  - Few queries from `semgrep_project` have keywords in upper case like `SELECT`, `FROM`, and `WHERE`  please can you change everything to lowercase to follow the standard
- `docs/index.md`
  - Please add a `.` at the end of the description
  - The output of the example query is missing the end line it should be added
  - Please format the query with `2 indent spaces` with [free formatted](https://www.freeformatter.com/sql-formatter.html#before-output)
  - Also please make sure that the `Configuration section` is in sync with the changes suggested for the config/semgrep.spc file.
- `semgrep/plugin.go`
  - The indentation is incorrect please use Ctrl + S or Cmd + S in case you are using a code editor like VS code
  - The list of tables should be sorted in ascending order
- `semgrep/utils.go`
  - We should add the configuration for loading the connection from the cache, which ensures throttling protection, etc. 
  -  Therefore we should also add the configuration to set it back to the cache once the connection is successfully created.
- `semgrep/*`
  - The logger error messages still include the old table names that are in plural please update them to use the updated table names for instance `plugin.Logger(ctx).Error("semgrep_deployment.listDeployments", "connection_error", err)` instead of `plugin.Logger(ctx).Error("semgrep_deployments.listDeployments", "connection_error", err)` (Extend to all tables)
  - Please update the table function names to singular for instance `tableDeployment` instead of `tableDeployments` (Extend to all tables)
- `README`
  - Please update the community link to 
  `Community: [Join #steampipe on Slack →](https://turbot.com/community/join)`
  - The output of the example query is missing the end line it should be added
  - `Configure the api token` here `API` should be in caps

**Comment by gabrielsoltz** [comment]
Hi @madhushreeray30, thank you for reviewing it again, 

I applied all the suggestions here #11 

What I don't understand is why you are suggesting correcting the indentation for `semgrep/plugin.go`, I'm using `go fmt` and the indentation is correct, let me know if I'm missing something. 

Thank you again,

**Comment by madhushreeray30** [comment]
@gabrielsoltz thanks for addressing all the issues the plugin code looks good to me 👍
